nntrainer/tensor/swap_device.h

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2022 Jiho Chu <jiho.chu@samsung.com>
+ *
+ * @file   swap_device.cpp
+ * @date   01 July 2022
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jiho Chu <jiho.chu@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  Swap device class
+ *
+ */
+
+#ifndef __SWAP_DEVICE_H__
+#define __SWAP_DEVICE_H__
+
+#include <fcntl.h>
+#include <map>
+#include <memory>
+#include <string>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <system_error>
+#include <unistd.h>
+#include <utility>
+
+
+/* Uncomment this to use mmap for swap data */
+//#define USE_MMAP
+
+namespace nntrainer {
+
+/**
+ * @class   SwapDevice
+ * @brief   A device used to storing data with long access time
+ */
+class SwapDevice {
+public:
+  /**
+   * @brief swap device default path
+   *
+   */
+  const std::string swap_device_path = "/tmp/";
+
+  /**
+   * @brief SwapDevice default constructor
+   *
+   */
+  explicit SwapDevice(const std::string &name)
+    : dev_path(swap_device_path + name), fd(-1) {}
+
+  /**
+   * @brief SwapDevice destructor
+   *
+   */
+  virtual ~SwapDevice() = default;
+
+  /**
+   * @brief Start device
+   *
+   * @param size The size of requested swap device space
+   *
+   */
+  void start(size_t size);
+
+  /**
+   * @brief Allocate and get data
+   *
+   * @param offset Requested offset of swap device file
+   * @param size Requested size.
+   *
+   * @return The pointer of the swap space
+   *
+   */
+  void *getBuffer(int offset, size_t size);
+
+  /**
+   * @brief Deallocate and put data
+   *
+   * @param ptr The pointer obtained from getBuffer
+   *
+   */
+  void putBuffer(void *ptr);
+
+  /**
+   * @brief Close device
+   *
+   */
+  void finish();
+
+  /**
+   * @brief Check device is operating
+   *
+   * @return Device operating status
+   *
+   */
+  bool isOperating() const { return (fd >= 0); }
+
+  /**
+   * @brief Get device path
+   *
+   * @return Device path
+   *
+   */
+  const std::string getDevicePath() const { return dev_path; }
+
+private:
+  const std::string dev_path; /**< device path */
+  int fd;                     /**< device file description */
```

Comment:
I'm not sure but do you think it might be better to use multiple fd w.r.t performance. Cause.. there might be a case which using swap device over 1GB memory. So.. if we knew the offset and get and put using multiple fd, then I think it could be better. Let me know your thought.

Reply:
Performance comparision is needed for it.
I'm preparing IO perf measure, so let's discuss more with the benchmark result.